ios/RNSScreen.mm

```diff
@@ -980,6 +1031,81 @@ - (void)viewDidLayoutSubviews
   }
 }
 
+// Checks whether this screen has any child view controllers of type RNSNavigationController.
+// Useful for checking if this screen has nested stack or is displayed at the top.
+- (BOOL)hasNestedStack
+{
+  for (UIViewController *vc in self.childViewControllers) {
+    if ([vc isKindOfClass:[RNSNavigationController class]]) {
+      return YES;
+    }
+  }
+
+  return NO;
+}
+
+- (CGFloat)getCalculatedHeaderHeightIsModal:(BOOL)isModal
+{
+  CGFloat navbarHeight = self.navigationController.navigationBar.frame.size.height;
+
+  // In case where screen is a modal, we want to calculate just its childViewController's height
+  if (isModal && self.childViewControllers.count > 0 &&
+      [self.childViewControllers[0] isKindOfClass:UINavigationController.class]) {
+    UINavigationController *childNavCtr = self.childViewControllers[0];
+    navbarHeight = childNavCtr.navigationBar.frame.size.height;
+  }
+
+  return navbarHeight;
+}
+
+- (CGSize)getCalculatedStatusBarHeightIsModal:(BOOL)isModal
+{
+#if !TARGET_OS_TV
+  BOOL isDraggableModal = isModal && ![self.screenView isFullscreenModal];
+  BOOL isDraggableModalWithChildViewCtr =
+      isDraggableModal && self.childViewControllers.count > 0 && self.childViewControllers[0] != nil;
+
+  // When modal is floating (we can grab its header), we don't want to calculate status bar in it.
+  // Thus, we return '0' as a height of status bar.
+  if (isDraggableModalWithChildViewCtr || self.screenView.isTransparentModal) {
+    return CGSizeMake(0, 0);
+  }
+
+  CGSize fallbackStatusBarSize = [[UIApplication sharedApplication] statusBarFrame].size;
+
+#if defined(__IPHONE_OS_VERSION_MAX_ALLOWED) && defined(__IPHONE_13_0) && \
+    __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_13_0
+  if (@available(iOS 13.0, *)) {
+    CGSize primaryStatusBarSize = self.view.window.windowScene.statusBarManager.statusBarFrame.size;
+    if (primaryStatusBarSize.height == 0 || primaryStatusBarSize.width == 0)
+      return fallbackStatusBarSize;
+
+    return primaryStatusBarSize;
+  } else {
+    return fallbackStatusBarSize;
+  }
+#endif /* Check for iOS 13.0 */
+
+#else
+  // On TVOS, status bar doesn't exist
+  return CGSizeMake(0, 0);
+#endif // !TARGET_OS_TV
+}
+
+- (void)calculateHeaderHeightIsModal:(BOOL)isModal
+{
+  CGFloat navbarHeight = [self getCalculatedHeaderHeightIsModal:isModal];
+  CGSize statusBarSize = [self getCalculatedStatusBarHeightIsModal:isModal];
+
+  // Unfortunately, UIKit doesn't care about switching width and height options on screen rotation.
+  // We should check if user has rotated its screen, so we're choosing the minimum value between the
+  // width and height.
+  CGFloat statusBarHeight = MIN(statusBarSize.width, statusBarSize.height);
+  CGFloat totalHeight = navbarHeight + statusBarHeight;
+
+  [self.screenView notifyHeaderHeightChange:totalHeight];
+}
```

Comment:
Do we want to dispatch event to JS on each call of this method? 
I would suggest splitting it into two separate methods:

1. actually resolving header height and returning it

```objective-c
- (CGFloat)calculateHeaderHeightIsModal:(BOOL)isModal
{ ... }
```
2. sending the event

Or in minimal version rename the method so that it is **clear** that it produces side-effects.

Reply:
Changed in https://github.com/software-mansion/react-native-screens/pull/1802/commits/aabfc7edf48d5521de9d6e4b34a6fdceef8e0e85.